src/libpsl-native/src/createprocess.cpp

```diff
@@ -0,0 +1,197 @@
+#include "createprocess.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <limits>
+#include <limits.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <sys/resource.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <syslog.h>
+#include <unistd.h>
+
+enum
+{
+    SUPPRESS_PROCESS_SIGINT = 0x00000001
+};
+
+enum
+{
+    READ_END_OF_PIPE = 0,
+    WRITE_END_OF_PIPE = 1
+};
+
+static void CloseIfOpen(int fd)
+{
+    if (fd >= 0)
+    {
+        close(fd); // Ignoring errors from close is a deliberate choice
+    }
+}
+
+// Checks if the IO operation was interrupted and needs to be retried.
+// Returns true if the operation was interrupted; otherwise, false.
+template <typename TInt>
+static inline bool CheckInterrupted(TInt result)
+{
+    return result < 0 && errno == EINTR;
+}
+
+static int Dup2WithInterruptedRetry(int oldfd, int newfd)
+{
+    int result;
+    while (CheckInterrupted(result = dup2(oldfd, newfd)));
+    return result;
+}
+
+int32_t ForkAndExecProcess(
+    const char* filename,
+    char* const argv[],
+    char* const envp[],
+    const char* cwd,
+    int32_t redirectStdin,
+    int32_t redirectStdout,
+    int32_t redirectStderr,
+    int32_t creationFlags,
+    int32_t* childPid,
+    int32_t* stdinFd,
+    int32_t* stdoutFd,
+    int32_t* stderrFd)
+{
+    int success = true;
+    int processId = -1;
+    int stdinFds[2] = { -1, -1 };
+    int stdoutFds[2] = { -1, -1 };
+    int stderrFds[2] = { -1, -1 };
+
+    // Validate arguments
+    if (nullptr == filename || nullptr == argv || nullptr == envp || nullptr == stdinFd || nullptr == stdoutFd ||
+        nullptr == stderrFd || nullptr == childPid)
+    {
+        assert(false && "null argument.");
+        errno = EINVAL;
+        success = false;
+        goto done;
+    }
+
+    if ((redirectStdin & ~1) != 0 || (redirectStdout & ~1) != 0 || (redirectStderr & ~1) != 0)
+    {
+        assert(false && "Boolean redirect* inputs must be 0 or 1.");
+        errno = EINVAL;
+        success = false;
+        goto done;
+    }
+
+    // Make sure we can find and access the executable. exec will do this, of course, but at that point it's already
+    // in the child process, at which point it'll translate to the child process' exit code rather than to failing
+    // the Start itself.  There's a race condition here, in that this could change prior to exec's checks, but there's
+    // little we can do about that. There are also more rigorous checks exec does, such as validating the executable
+    // format of the target; such errors will emerge via the child process' exit code.
+    if (access(filename, X_OK) != 0)
+    {
+        success = false;
+        goto done;
+    }
+
+    // Open pipes for any requests to redirect stdin/stdout/stderr
+    if ((redirectStdin && pipe(stdinFds) != 0) || (redirectStdout && pipe(stdoutFds) != 0) ||
+        (redirectStderr && pipe(stderrFds) != 0))

```

Comment:
Without `close-on-exec`, the pipe file descriptors will be leaked to all subsequent processes created by powershell. The symptom I described above is one result this would cause. Although it's not applicable to the SSH remoting scenario, my worry is that there may be other symptoms resulted by the leak that would cause a problem we don't know yet which might be hard to diagnose.